### PR TITLE
Overhaul logic to synchronize krnlmon with stratum

### PR DESCRIPTION
**Pull request by ffoulkes**
1) Replaced global pthreads variables with absl::Notification
   objects that are instantiated by infrap4d_main and passed
   as parameters.

2) Implemented target_main() to abstract the call to the main
   function for the selected target.

3) Removed nominal support for TOFINO_TARGET until such time
   as it is needed.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

### CODE REVIEW

File: infrap4d/infrap4d_main.cc [hunk 1]
```diff
@@ -2,48 +2,51 @@
...
       daemonize_complete();
   }
 
-  krnlmon_init();
+  absl::Notification ready_sync;
+  absl::Notification done_sync;
 
-  krnlmon_shutdown();
+  krnlmon_create_main_thread(&ready_sync);
+  krnlmon_create_shutdown_thread(&done_sync);
```
Comment (n-sandeep) [inline infrap4d/infrap4d_main.cc:43]: Why do we need this shutdown to be called immediately after creating a switchlink thread?

Comment (ffoulkes) [reply infrap4d/infrap4d_main.cc:43]: This function doesn't shut down; it creates a thread that _waits_ for a signal indicating that DpdkHal has shut down. That's why I renamed the function called here from "shutdown" to "create shutdown thread."

Unless I'm greatly mistaken, the existing shutdown thread has a bug in it. I'm pretty sure we should be passing `main_tid` to `switchlink_stop` so the latter can use the pid to stop the main switchlink thread. (Instead, it's using an uninitialized variable as the thread ID.)

I elected not to fix this in the initial Draft PR in order to focus on using dependency inversion to break the tight coupling between DpdkHal and Krnlmon.

**Review by nupuruttarwar — COMMENTED** [review]
Conversion from pthread mechanism to abseil Notification LGTM, we will need to test this. @n-sandeep, can you please help with testing

**Comment by n-sandeep** [comment]
> Conversion from pthread mechanism to abseil Notification LGTM, we will need to test this. @n-sandeep, can you please help with testing

@nupuruttarwar, Let me discuss with Derek and take it forward.

File: infrap4d/infrap4d_main.cc [hunk 2]
```diff
@@ -2,48 +2,51 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #if defined(DPDK_TARGET)
-#include "stratum/hal/bin/tdi/dpdk/dpdk_main.h"
-#elif defined (TOFINO_TARGET)
```
Comment (5abeel) [inline infrap4d/infrap4d_main.cc:6]: Why are we getting rid of the TOFINO_TARGET from here?

**Pull request closed without merging**